Mathlib/Tactic/Linter/DependentTypeclass.lean

```diff
@@ -0,0 +1,121 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Lean
+import Mathlib.Lean.Expr.Basic
+
+/-!
+#  The "dependentTypeclass" linter
+
+The "dependentTypeclass" linter emits a warning when there is a dependency among typeclass
+assumptions.
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter
+
+/-- The "dependentTypeclass" linter emits a warning when there is a dependency among typeclass
+assumptions.
+ -/
+register_option linter.dependentTypeclass : Bool := {
+  defValue := true
+  descr := "enable the dependentTypeclass linter"
+}
+def getBinders : Expr → Array String
+  | .lam na type body bi =>
+    let (l, r) := bi.brackets
+    (getBinders body).push (s!"{l} {na} : {type}{r}")
+  | .forallE na type body bi =>
+    let (l, r) := bi.brackets
+    (getBinders body).push (s!"{l} {na} : {type}{r}")
+  | _ => default
+
+/-- `getMinInstsFor d varDecls` takes as input a `bracketedBinder` `d` and an array of `varDecls`
+of ``bracketedBinder`.
+It assumes that `d` represents a typeclass assumption and it tries to synthesize an instance of
+`d` from `varDecls` using `inferInstance` or `sorry`.
+If `inferInstance` is successful, then it returns the `Expr`ession representing a proof of
+`d` using `varDecls`.
+Otherwise it returns `none`.
+-/
+def getMinInstsFor (d : TSyntax ``Lean.Parser.Term.bracketedBinder)
+    (varDecls : Array (TSyntax ``Lean.Parser.Term.bracketedBinder)) :
+      CommandElabM (Option Expr) := do
+  let sc ← getScope
+  let newName ← liftCoreM <| mkFreshUserName `hello
+  let stxId ← `(declId| $(mkIdent newName))
+  let tac ← `(tacticSeq| first | infer_instance | sorry)
```

Comment:
This seems like a very large hammer for the task of finding instances. 

What do you want to run continuously versus in CI? 

Looking at the Zulip conversation, I would guess, for running continuously, it would be enough if we look at `LocalInstances` and check for overlapping projections to some data field. 

For CI, I think we can have a flag that is more aggressive so that people don't do `variable [Field \R]` for example. 


What do you think?